src/OrchardCore.Modules/OrchardCore.Rules/Services/RuleService.cs

```diff
@@ -0,0 +1,53 @@
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace OrchardCore.Rules.Services
+{
+    public class RuleService : IRuleService
+    {
+        private static readonly ValueTask<bool> False = new ValueTask<bool>(false);
+        private static readonly ValueTask<bool> True = new ValueTask<bool>(true);
+
+        private readonly IConditionResolver _conditionResolver;
+
+        public RuleService(IConditionResolver conditionResolver)
+        {
+            _conditionResolver = conditionResolver;
+        }
+
+        public ValueTask<bool> EvaluateAsync(Rule rule)
+        {          
+            static async ValueTask<bool> Awaited(ValueTask<bool> task)
+                => await task;
+            
+            foreach(var childCondition in rule.Conditions)
+            {
+                var evaluator = _conditionResolver.GetConditionEvaluator(childCondition);
+
+                var task = evaluator.EvaluateAsync(childCondition);
+                if (!task.IsCompletedSuccessfully)
+                {
+                    if (!Awaited(task).Result)
```

Comment:
I understand the pattern but because you can't return directly if true, here you are  forced to run it synchronously, is it intended, even if most of the time it completes immediately? Here it's like a `Getawaiter().GetResult()`

Reply:
Reverted @jtkech - I don't well understand that pattern, so was hoping for better optimization.